File: src/main/org/nlogo/prim/etc/_abs.java
// (C) 2012 Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc;

import org.nlogo.api.LogoException;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Pure;
import org.nlogo.nvm.Reporter;

public final strictfp class _abs extends Reporter implements Pure {
  @Override
  public Syntax syntax() {
    return Syntax.reporterSyntax
        (new int[]{Syntax.NumberType()},
            Syntax.NumberType());
  }

  @Override
  public Object report(Context context) throws LogoException {
    return report_1(context, argEvalDoubleValue(context, 0));
  }

  public Double report_1(Context context, Double d) {
    double unwrapped = d.doubleValue();
    return (unwrapped < 0) ? -unwrapped : d;
  }

  public double report_2(Context context, double d) {
    return (d < 0) ? -d : d;
  }
}

File: src/main/org/nlogo/prim/etc/_abs.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc

import org.nlogo.api.Syntax
import org.nlogo.nvm.{ Context, Pure, Reporter }

@annotation.strictfp
class _abs extends Reporter with Pure {

  override def syntax =
    Syntax.reporterSyntax(
      Array(Syntax.NumberType), Syntax.NumberType)

  override def report(context: Context): java.lang.Double =
    report_1(context, argEvalDoubleValue(context, 0))

  def report_1(context: Context, d: java.lang.Double): java.lang.Double = {
    val unwrapped = d.doubleValue
    if (unwrapped < 0)
      Double.box(-unwrapped)
    else
      d
  }

  def report_2(context: Context, d: Double): Double =
    if (d < 0) -d else d

}
